neard/src/log_config_watcher.rs

```diff
@@ -0,0 +1,54 @@
+use near_o11y::reload_env_filter;
+use serde::{Deserialize, Serialize};
+use std::path::PathBuf;
+use tracing::{error, info};
+
+/// Configures logging.
+#[derive(Serialize, Deserialize, Clone, Debug)]
+struct LogConfig {
+    /// Comma-separated list of EnvFitler directives.
+    pub rust_log: Option<String>,
+    /// Some("") enables global debug logging.
+    /// Some("module") enables debug logging for "module".
+    pub verbose_module: Option<String>,
+}
+
+pub(crate) struct LogConfigWatcher {
+    pub watched_path: PathBuf,
+}
+
+pub(crate) enum UpdateBehavior {
+    UpdateOrReset,
+    UpdateOnlyIfExists,
+}
+
+impl LogConfigWatcher {
+    pub fn update(&self, update_behavior: UpdateBehavior) {
+        match std::fs::read_to_string(&self.watched_path) {
+            Ok(log_config_str) => match serde_json::from_str::<LogConfig>(&log_config_str) {
+                Ok(log_config) => {
+                    info!(target: "neard", log_config=?log_config, "Changing the logging config.");
+                    if let Err(err) = reload_env_filter(
+                        log_config.rust_log.as_deref(),
+                        log_config.verbose_module.as_deref(),
+                    ) {
+                        error!(target: "neard", err=?err, "Failed to reload the logging config.");
+                    }
+                    return;
+                }
+                Err(err) => {
+                    error!(target: "neard", logging_config_path=%self.watched_path.display(), err=?err, "Ignoring the logging config change because failed to parse logging config file.");
+                    return;
+                }
+            },
+            Err(err) => {
+                if let UpdateBehavior::UpdateOrReset = update_behavior {
+                    info!(target: "neard", logging_config_path=%self.watched_path.display(), err=?err, "Reset the logging config because the logging config file doesn't exist.");
+                    if let Err(err) = reload_env_filter(None, None) {
+                        error!(target: "neard", err=?err, "Failed to reload the logging config");
+                    }
+                }
+            }
+        }
+    }
```

Comment:
To me this screams for a helper function with fewer indentation levels and single place where the error is printed:

```rust
impl LogConfigWatcher {
    pub fn update(&self, update_behavior: UpdateBehavior) {
        if let Err(err) = self.do_update(update_behavior) {
            error!(target: "neard", logging_config_path=%self.watched_path.display(), err=?err, "Failed to reload the logging config.");
        }
    }

    fn do_update(&self) -> anyhow::Result<()> {
        let update_or_reset = matches!(update_behavior, UpdateBehavior::UpdateOrReset);
        let data = match std::fs::read_to_string(&self.watched_path) {
            Ok(data) => data,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound && update_or_reset => {
                info!(target: "neard", "Resetting the logging config.");
                reload_env_filter(None, None)?;
                return Ok(());
            },
            Err(err) => {
                return Err(err.into());
            }
        };

        let data = serde_json::from_str::<LogConfig>(&log_config_str)?;
        info!(target: "neard", logging_config_path=%self.watched_path.display(), "Updating the logging config.");
        reload_env_filter(
            log_config.rust_log.as_deref(),
            log_config.verbose_module.as_deref(),
        )?;
        Ok(())
    }
}
```

Reply:
Reworked error handling, thanks for the suggestion.